spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/config/AbstractSpannerConfiguration.java

```diff
@@ -0,0 +1,106 @@
+/*
+ *  Copyright 2018 original author or authors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.data.spanner.config;
+
+import com.google.auth.Credentials;
+import com.google.cloud.spanner.DatabaseClient;
+import com.google.cloud.spanner.DatabaseId;
+import com.google.cloud.spanner.Spanner;
+import com.google.cloud.spanner.SpannerOptions;
+
+import org.springframework.cloud.gcp.data.spanner.core.SpannerMutationFactory;
+import org.springframework.cloud.gcp.data.spanner.core.SpannerMutationFactoryImpl;
+import org.springframework.cloud.gcp.data.spanner.core.SpannerOperations;
+import org.springframework.cloud.gcp.data.spanner.core.SpannerTemplate;
+import org.springframework.cloud.gcp.data.spanner.core.convert.MappingSpannerConverter;
+import org.springframework.cloud.gcp.data.spanner.core.convert.SpannerConverter;
+import org.springframework.cloud.gcp.data.spanner.core.mapping.SpannerMappingContext;
+import org.springframework.context.annotation.Bean;
+
+/**
+ * @author Balint Pato
+ */
+public abstract class AbstractSpannerConfiguration {
```

Comment:
Is there a reason we're not using `@ConditionalOnMissingBean` on `GcpSpannerAutoConfiguration` and then just overriding the beans we need for the tests?

Reply:
I didn't want to depend on Spring Boot for integration testing. Don't you think that would be an overkill?

Reply:
We don't have to depend on Spring Boot for testing. We can inject beans into the context in other ways.
But it sounds like we're adding this `AbstractSpannerConfiguration` class because of integration tests, and I wouldn't want to add it if we don't need it.

Reply:
The `GcpSpannerAutoConfiguration` is already Spring Boot based, so, I agree with @joaoandremartins : it would be better to mark beans with the `ConditionalOnMissingBean` to let end-user to override them in his/her target application.
Or even for us in the test suit.

It indeed smells to introduce something in the `main` what purpose is only for testing.